internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,281 @@
+		}
+	}
+
+	// wildcard should have precedence
+	if wb.Len() > 0 {
+		b.WriteString(wb.String())
+	}
+
+	if rb.Len() > 0 {
+		b.WriteString(rb.String())
+	}
+
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+func (c *CachedDatastore) Read(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options storage.ReadOptions) (storage.TupleIterator, error) {
+	ctx, span := tracer.Start(ctx, "cache.Read", trace.WithAttributes(attribute.Bool("cached", false)))
+	defer span.End()
+
+	iter := func(ctx context.Context) (storage.TupleIterator, error) {
+		return c.OpenFGADatastore.Read(ctx, store, tupleKey, options)
+	}
+
+	if options.Consistency.Preference == openfgav1.ConsistencyPreference_HIGHER_CONSISTENCY {
+		return iter(ctx)
+	}
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("%sr%s/%s", QueryCachePrefix, store, tuple.TupleKeyToString(tupleKey)))
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+type iterFunc func(ctx context.Context) (storage.TupleIterator, error)
+
+func (c *CachedDatastore) newCachedIterator(ctx context.Context, dsIterFunc iterFunc, key string) (storage.TupleIterator, error) {
+	span := trace.SpanFromContext(ctx)
+	tuplesCacheTotalCounter.Inc()
+
+	cacheKey := cacheKeyFor(key)
+
+	if cacheKey == "" {
+		return dsIterFunc(ctx)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter          storage.TupleIterator
+	tuples        []*openfgav1.Tuple
+	cacheKey      string
+	cache         storage.InMemoryCache[any]
+	ttl           time.Duration
+	maxResultSize int
+	closeOnce     sync.Once
+	wg            sync.WaitGroup
+	sf            singleflight.Group
+}
+
+func (c *cachedIterator) addToBuffer(t *openfgav1.Tuple) bool {
+	if c.tuples == nil {
+		return false
+	}
+	c.tuples = append(c.tuples, t)
+	if len(c.tuples) >= c.maxResultSize {
+		tuplesCacheDiscardCounter.Inc()
+		c.tuples = nil
+	}
+	return true
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) && !errors.Is(err, context.Canceled) && !errors.Is(err, context.DeadlineExceeded) {
+			c.tuples = nil // don't store results that are incomplete
```

Comment:
Question: why do we want to store in the case of context.canceled or context.deadlineexceeded?  These are error cases where we don't have the complete result.  IMO, we probably want to be on the safe side and discard the cached result.

Reply:
If the request resolves before finishing the iterator, it will receive cancelled event from the upstream HTTP context. This will happen to the majority of iterators, but we still want to cache the in the background. It won't be a partial result as on `.Close` we continue draining with a different context, not the parent associated with the Check (in this case) resolution.

Reply:
In this case, we rely on an implicit behaviour that Next() MUST NOT advance / discard data upon context timeout/deadline.  Perhaps we need to document this expectation in the iterator interface?

Another possibility is to check for context deadline/cancel prior to line 204.  This will avoid the implicit behaviour dependency.

Reply:
On a side note, we should unit test this line (i.e., next yielding error) - we can use mock iterator to do that.

Reply:
> On a side note, we should unit test this line (i.e., next yielding error) - we can use mock iterator to do that.

added this test